query_executor.go

```diff
@@ -201,16 +210,375 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
 	}
 
 	if lastErr != nil {
-		return &Iter{err: lastErr}
+		return newErrIter(lastErr, qry.getQueryMetrics())
 	}
 
-	return &Iter{err: ErrNoConnections}
+	return newErrIter(ErrNoConnections, qry.getQueryMetrics())
 }
 
-func (q *queryExecutor) run(ctx context.Context, qry ExecutableQuery, hostIter NextHost, results chan<- *Iter) {
+func (q *queryExecutor) run(ctx context.Context, qry internalRequest, hostIter NextHost, results chan<- *Iter) {
 	select {
 	case results <- q.do(ctx, qry, hostIter):
 	case <-ctx.Done():
 	}
-	qry.releaseAfterExecution()
+}
+
+type queryOptions struct {
+	stmt                  string
+	values                []interface{}
+	initialConsistency    Consistency
+	pageSize              int
+	initialPageState      []byte
+	prefetch              float64
+	trace                 Tracer
+	observer              QueryObserver
+	rt                    RetryPolicy
+	spec                  SpeculativeExecutionPolicy
+	binding               func(q *QueryInfo) ([]interface{}, error)
+	serialCons            SerialConsistency
+	defaultTimestamp      bool
+	defaultTimestampValue int64
+	disableSkipMetadata   bool
+	context               context.Context
+	idempotent            bool
+	customPayload         map[string][]byte
+	keyspace              string
+	disableAutoPage       bool
+	skipPrepare           bool
+	routingKey            []byte
+
+	// getKeyspace is field so that it can be overriden in tests
+	getKeyspace func() string
+}
+
+func newQueryOptions(q *Query) *queryOptions {
+	return &queryOptions{
```

Comment:
Should we be copying `values`, `customPayload`, `initialPageState`, `routingKey`? it might not be obvious that those are not copied and can't be mutated.

Reply:
I think copying `values` and `customPayload` might be a bit too much due to performance reasons. `initialPageState` should probably be copied for safety and since its length is usually pretty small. `routingKey` I'm not so sure, I think it will be `nil` the vast majority of cases anyway but I can make sure it gets copied as well